w3af/plugins/audit/hpp.py

```diff
@@ -0,0 +1,259 @@
+Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+"""
+
+from io import StringIO
+from w3af.core.controllers.plugins.audit_plugin import AuditPlugin
+from w3af.core.data.fuzzer.fuzzer import create_mutants
+from w3af.core.data.kb.vuln import Vuln
+from w3af.core.data.kb.info import Info
+from w3af.core.data.fuzzer.mutants.querystring_mutant import QSMutant
+import w3af.core.controllers.output_manager as om
+import w3af.core.data.constants.severity as severity
+from w3af.core.controllers.misc.fuzzy_string_cmp import fuzzy_equal
+import w3af.core.data.parsers.parser_cache as parser_cache
+from lxml import etree
+from w3af.core.controllers.exceptions import BaseFrameworkException
+from w3af.core.data.parsers.url import URL
+from w3af.core.data.kb.info_set import InfoSet
+
+
+REFLECT_TOKEN = 'w3afHPPReflectToken'
+FORGE_TOKEN = 'w3afHPPForgeToken'
+EQUALITY_THRESHOLD = 0.95
+#paramater precedence types
+LAST = 'last occurance'
+FIRST = 'first occurance'
+CONCAT = 'concatination'
+UNK = 'unknown'
+
+
+class hpp(AuditPlugin):
+    """
+    Identify http parameter pollution vulnerabilities.
+    """
+
+    def __init__(self):
+        AuditPlugin.__init__(self)
+        self.parser = etree.HTMLParser(remove_blank_text=True, remove_comments=True)
+        self.precedence = None
+        self.rlink = ''
+
+
+    def audit(self, freq, orig_response):
+        """
+        Tests an URL for HPP vulnerabilities.
+
+        :param freq: A FuzzableRequest
+
+        """
+
+        rparams = {}
+
+        if not orig_response.is_text_or_html():
+            return
+
+        if not orig_response.get_code() == 200:
+            return
+        method = freq.get_method()
+        self._orig_uri = freq.get_uri()
+        if method == 'GET':
+            qs = freq.get_querystring()
+
+
+        #finding reflected params
+        for param in qs:
+            value = qs[param][0]
+            mutants = create_mutants(freq, [REFLECT_TOKEN],
+                                     fuzzable_param_list= [param],
+                                     mutant_tuple=(QSMutant,))
+            mu = mutants[0]
+            #first we send REFLECT_TOKEN
+            resp = self._uri_opener.send_mutant(mu)
+            try:
+                dp = parser_cache.dpc.get_document_parser_for(resp)
+            except BaseFrameworkException:
+                return
+            #do we have a reflection?
+
+            body = self.strip_trash(resp.body)
+            if not resp.get_code() == 200:
+                #no need to seek reflection if it is not OK
+                continue
+            rp = self.find_reflection( dp, REFLECT_TOKEN)
+
+            if not rp:
+                # no reflection at all skip the parameter
+                continue
+            rparams[param] = rp
+            # print "querystring parameter %s is reflected inside " \
+            #       "a link HTML code" % param
+            if self.precedence:
```

Comment:
So, if at any point in time we found/set a precedence then nothing else is done. There is something wrong with this logic, since:

 * Call #1 to audit() sends N requests and sets precedence
 * Call #2 to audit() sends M requests and no matter what they return, nothing is done.
 * Call #N to audit() sends Y requests and no matter what they return, nothing is done.

So we're wasting a lot of HTTP requests once the precedence is already set. I would put something like this at the top of the audit method code:

```
if self.precedence:
    return
```